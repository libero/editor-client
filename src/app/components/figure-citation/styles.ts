import { makeStyles } from '@material-ui/core/styles';

export const useFigureCitationEditorStyles = makeStyles((theme) => ({
  figureSelectionList: {
    width: 380,
    maxHeight: 280,
    overflow: 'auto'
  },
  figureSelectionListItem: {
    cursor: 'pointer',
    padding: theme.spacing(1, 1.5),
    display: 'flex',
    margin: theme.spacing(0, 1.5),
    border: 'solid 1px rgba(0, 0, 0, 0.23)',
    borderBottom: 'none',
    '&:first-child': {
      borderTopLeftRadius: theme.spacing(0.5),
      borderTopRightRadius: theme.spacing(0.5)
    },
    '&:last-child': {
      borderBottom: 'solid 1px rgba(0, 0, 0, 0.23)',
      marginBottom: theme.spacing(1.5),
      borderBottomLeftRadius: theme.spacing(0.5),
      borderBottomRightRadius: theme.spacing(0.5)
    }
  },
  clearFilterIcon: {
    cursor: 'pointer'
  },
  filterField: {
    width: `calc(100% - ${theme.spacing(3)}px)`,
    margin: theme.spacing(1.5, 1.5, 1, 1.5),
    '& input': {
      padding: theme.spacing(1.25, 1.75)
    }
  },
  hiddenIcon: {
    visibility: 'hidden'
  },
  figureContent: {
    flexGrow: 1,
    paddingTop: theme.spacing(1)
  },
  figureTick: {
    alignSelf: 'center',
    flexGrow: 0
  }
}));
