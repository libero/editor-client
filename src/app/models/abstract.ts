import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { gapCursor } from 'prosemirror-gapcursor';

import { makeSchemaFromConfig } from 'app/models/utils';
import * as abstractConfig from 'app/models/config/abstract.config';
import { buildInputRules } from 'app/models/plugins/input-rules';
import { PlaceholderPlugin } from 'app/models/plugins/placeholder.plugin';
import { SelectionPlugin } from 'app/models/plugins/selection.plugins';

export function createAbstractState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), SelectionPlugin, gapCursor(), PlaceholderPlugin('Enter abstract')]
  });
}

export function createImpactStatementState(content: Element): EditorState {
  const schema = makeSchemaFromConfig(abstractConfig.topNode, abstractConfig.nodes, abstractConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), SelectionPlugin, gapCursor(), PlaceholderPlugin('Enter impact statement')]
  });
}
