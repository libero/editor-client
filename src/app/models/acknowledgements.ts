import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';
import { Step } from 'prosemirror-transform';
import { gapCursor } from 'prosemirror-gapcursor';

import * as acknowledgementsConfig from './config/acknowledgements.config';
import { buildInputRules } from './plugins/input-rules';
import { makeSchemaFromConfig } from './utils';
import { PlaceholderPlugin } from './plugins/placeholder.plugin';
import { SelectionPlugin } from './plugins/selection.plugins';

export function createAcknowledgementsState(content?: Element, changeSteps?: [Step]): EditorState {
  if (content) {
    const ackTitle = content.querySelector('title');
    if (ackTitle) {
      content.removeChild(ackTitle);
    }
  }

  const schema = makeSchemaFromConfig(
    acknowledgementsConfig.topNode,
    acknowledgementsConfig.nodes,
    acknowledgementsConfig.marks
  );
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [buildInputRules(), SelectionPlugin, gapCursor(), PlaceholderPlugin('Enter acknowledgements')]
  });
}
