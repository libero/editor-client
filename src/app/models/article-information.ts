import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { buildInputRules } from './plugins/input-rules';
import { gapCursor } from 'prosemirror-gapcursor';
import { dropCursor } from 'prosemirror-dropcursor';

import { getTextContentFromPath, makeSchemaFromConfig } from './utils';
import * as licenseTextConfig from './config/license-text.config';
import { Person } from './person';
import moment from 'moment';
import { BackmatterEntity } from './backmatter-entity';
import { JSONObject } from '../types/utility.types';

export const LICENSE_CC_BY_4 = 'CC-BY-4';
export const LICENSE_CC0 = 'CC0';

const LICENSE_CC_BY_4_TEXT = `This article is distributed under the terms of the 
    <ext-link ext-link-type="uri" xlink:href="http://creativecommons.org/licenses/by/4.0/">
        Creative Commons Attribution License</ext-link>, which permits unrestricted use and redistribution provided that the original author 
    and source are credited.`;

const LICENSE_CC0_TEXT = `This is an open-access article, free of all copyright, and may be freely reproduced, 
    distributed, transmitted, modified, built upon, or otherwise used by anyone for any lawful purpose.
    The work is made available under the
    <ext-link ext-link-type="uri" xlink:href="http://creativecommons.org/publicdomain/zero/1.0/">Creative Commons CC0 public domain dedication</ext-link>.`;

export class ArticleInformation extends BackmatterEntity {
  public static getLicenseText(licenseType: string): EditorState {
    const paragraph = document.createElement('p');
    paragraph.innerHTML = {
      [LICENSE_CC_BY_4]: LICENSE_CC_BY_4_TEXT,
      [LICENSE_CC0]: LICENSE_CC0_TEXT
    }[licenseType];

    return ArticleInformation.createLicenseEditorState(paragraph);
  }

  public static createLicenseEditorState(content?: Element): EditorState {
    const schema = makeSchemaFromConfig(licenseTextConfig.topNode, licenseTextConfig.nodes, licenseTextConfig.marks);
    const xmlContentDocument = document.implementation.createDocument('', '', null);

    if (content) {
      xmlContentDocument.appendChild(content);
    }

    return EditorState.create({
      doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
      schema,
      plugins: [buildInputRules(), gapCursor(), dropCursor()]
    });
  }

  articleType: string;
  dtd: string;
  articleDOI: string;
  publisherId: string;
  volume: string;
  elocationId: string;
  subjects: Array<string>;
  publicationDate: string;
  licenseType: string;
  copyrightStatement: string;
  licenseText: EditorState;

  constructor(data?: JSONObject | Element, authors?: Person[]) {
    super();
    this.createEntity(data);
    if (data instanceof Element && authors) {
      this.updateCopyrightStatementFromXml(data, authors);
    }
  }

  public clone(): ArticleInformation {
    const newArticleInformation = new ArticleInformation();
    newArticleInformation.articleType = this.articleType;
    newArticleInformation.dtd = this.dtd;
    newArticleInformation.articleDOI = this.articleDOI;
    newArticleInformation.elocationId = this.elocationId;
    newArticleInformation.volume = this.volume;
    newArticleInformation.publisherId = this.publisherId;
    newArticleInformation.subjects = [...this.subjects];
    newArticleInformation.copyrightStatement = this.copyrightStatement;
    newArticleInformation.licenseType = this.licenseType;
    newArticleInformation.publicationDate = this.publicationDate;
    newArticleInformation.licenseText = this.licenseText;

    return newArticleInformation;
  }

  public updateCopyrightStatement(authors: Person[]): void {
    if (this.licenseType === LICENSE_CC_BY_4) {
      const authorNamesSection =
        authors.length === 1
          ? authors[0].lastName
          : authors.length === 2
          ? `${authors[0].lastName} and ${authors[1].lastName}`
          : authors.length > 2
          ? `${authors[0].lastName} et al`
          : '';
      const date = moment(this.publicationDate);

      this.copyrightStatement = `Â© ${date.isValid() ? date.format('YYYY') : ''}, ${authorNamesSection}`;
    } else {
      this.copyrightStatement = '';
    }
  }

  protected fromXML(xmlNode: Element): void {
    this.articleType = getTextContentFromPath(xmlNode, 'article-meta subj-group[subj-group-type="heading"]');
    const articleEl = xmlNode.querySelector('article');
    this.dtd = articleEl ? articleEl.getAttribute('dtd-version') : '';
    this.articleDOI = getTextContentFromPath(xmlNode, 'article-meta article-id[pub-id-type="doi"]');
    this.elocationId = getTextContentFromPath(xmlNode, 'article-meta elocation-id');
    this.volume = getTextContentFromPath(xmlNode, 'article-meta volume');
    this.publisherId = getTextContentFromPath(xmlNode, 'article-meta article-id[pub-id-type="publisher-id"]');
    this.subjects = Array.from(xmlNode.querySelectorAll('subj-group[subj-group-type="major-subject"] subject')).map(
      (el: Element) => el.textContent
    );

    this.licenseType = this.getLicenseTypeFromXml(xmlNode);
    this.publicationDate = this.getPublicationDateFromXml(xmlNode);
    this.licenseText = ArticleInformation.createLicenseEditorState(
      xmlNode.querySelector('article-meta permissions license license-p')
    );
  }

  protected fromJSON(json: JSONObject): void {
    this.articleType = json.articleType as string;
    this.dtd = json.dtd as string;
    this.articleDOI = json.articleDOI as string;
    this.elocationId = json.elocationId as string;
    this.volume = json.volume as string;
    this.publisherId = json.publisherId as string;
    this.subjects = (json.subjects as string[]) || [];

    this.licenseType = json.licenseType as string;
    this.publicationDate = json.publicationDate as string;
    const emptyEditorState = ArticleInformation.createLicenseEditorState();
    this.licenseText = EditorState.fromJSON(
      {
        schema: emptyEditorState.schema,
        plugins: emptyEditorState.plugins
      },
      json.licenseText as JSONObject
    );
  }

  protected createBlank(): void {
    this.articleType = '';
    this.dtd = '';
    this.articleDOI = '';
    this.elocationId = '';
    this.volume = '';
    this.publisherId = '';
    this.subjects = [];

    this.licenseType = '';
    this.publicationDate = '';
    this.licenseText = ArticleInformation.createLicenseEditorState();
  }

  private getLicenseTypeFromXml(doc: Element): string {
    const licenseEl = doc.querySelector('article-meta permissions license');
    if (!licenseEl) {
      return '';
    }
    const href = licenseEl.getAttribute('xlink:href');
    if (href === 'http://creativecommons.org/licenses/by/4.0/') {
      return LICENSE_CC_BY_4;
    }
    if (href === 'http://creativecommons.org/publicdomain/zero/1.0/') {
      return LICENSE_CC0;
    }
  }

  private getPublicationDateFromXml(xmlNode: Element): string {
    let publicationDate = '';
    const pubDateNode = xmlNode.querySelector('pub-date[date-type="pub"][publication-format="electronic"]');
    if (pubDateNode) {
      publicationDate = [
        pubDateNode.querySelector('year')?.textContent,
        pubDateNode.querySelector('month')?.textContent,
        pubDateNode.querySelector('day')?.textContent
      ].join('-');
    }

    return publicationDate;
  }

  private updateCopyrightStatementFromXml(xmlNode: Element, authors: Person[]): void {
    const copyrightStatementFromXml = getTextContentFromPath(xmlNode, 'article-meta permissions copyright-statement');
    if (copyrightStatementFromXml) {
      this.copyrightStatement = copyrightStatementFromXml;
    } else {
      this.updateCopyrightStatement(authors);
    }
  }
}
