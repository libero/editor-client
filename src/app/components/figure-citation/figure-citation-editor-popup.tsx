import React, { useCallback, useState, SyntheticEvent } from 'react';
import CheckCircleIcon from '@material-ui/icons/CheckCircle';
import { Popper, Paper, ClickAwayListener, TextField, InputAdornment } from '@material-ui/core';
import ClearIcon from '@material-ui/icons/Clear';

import { useFigureCitationEditorStyles } from 'app/components/figure-citation/styles';

export interface FiguresListEntry {
  id: string;
  name: string;
}

interface FigureCitationEditorPopupProps {
  selectedIds: string[];
  figures: FiguresListEntry[];
  onClose(): void;
  anchorEl: HTMLElement;
  onChange(ids: string[]): void;
}

export const FigureCitationEditorPopup: React.FC<FigureCitationEditorPopupProps> = (props) => {
  const { onClose, onChange, anchorEl, figures, selectedIds } = props;
  const [filteredList, setFilteredList] = useState<FiguresListEntry[]>(figures);
  const [filterValue, setFilterValue] = useState<string>('');
  const [internalSelectedIds, setInternalSelectedIds] = useState<string[]>(selectedIds);
  const classes = useFigureCitationEditorStyles();

  const handleFilterChange = useCallback(
    (event) => {
      const filterValue = event.target['value'].toLowerCase();
      setFilterValue(event.target['value']);
      setFilteredList(
        figures.filter((fig) => {
          return fig.name.toLowerCase().indexOf(filterValue) >= 0 || internalSelectedIds.includes(fig.id);
        })
      );
    },
    [internalSelectedIds, setFilterValue, figures]
  );
  const handleClick = useCallback(
    (event: SyntheticEvent) => {
      event.preventDefault();
      event.stopPropagation();
      const figId = (event.currentTarget as HTMLLIElement).dataset['figId'];
      const updatedIdsList = internalSelectedIds.includes(figId)
        ? internalSelectedIds.filter((id) => id !== figId)
        : [...internalSelectedIds, figId];
      setInternalSelectedIds(updatedIdsList);
      onChange(updatedIdsList);
    },
    [onChange, internalSelectedIds, setInternalSelectedIds]
  );

  const clearFilterField = useCallback(() => {
    setFilterValue('');
    setFilteredList(figures);
  }, [setFilterValue, setFilteredList, figures]);

  return (
    <Popper open={true} anchorEl={anchorEl}>
      <Paper>
        <ClickAwayListener onClickAway={onClose} mouseEvent="onMouseUp">
          <div>
            <TextField
              fullWidth
              classes={{ root: classes.filterField }}
              InputLabelProps={{ shrink: true }}
              label="Filter list"
              value={filterValue}
              variant="outlined"
              InputProps={{
                endAdornment: (
                  <InputAdornment position="end">
                    <ClearIcon
                      onClick={clearFilterField}
                      color="disabled"
                      classes={{ root: classes.clearFilterIcon }}
                    />
                  </InputAdornment>
                )
              }}
              onChange={handleFilterChange}
            />
            <ul className={classes.figureSelectionList}>
              {filteredList.map((fig) => (
                <li className={classes.figureSelectionListItem} key={fig.id} data-fig-id={fig.id} onClick={handleClick}>
                  <div className={classes.figureContent}>{fig.name}</div>
                  <div className={classes.figureTick}>
                    <CheckCircleIcon
                      color="primary"
                      className={!internalSelectedIds.includes(fig.id) ? classes.hiddenIcon : ''}
                    />
                  </div>
                </li>
              ))}
            </ul>
          </div>
        </ClickAwayListener>
      </Paper>
    </Popper>
  );
};
