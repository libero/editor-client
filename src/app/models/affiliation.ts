import { getTextContentFromPath } from 'app/models/utils';
import { BackmatterEntity } from 'app/models/backmatter-entity';
import { JSONObject } from 'app/types/utility.types';

export class Affiliation extends BackmatterEntity {
  label: string;
  institution: { name: string };
  address: { city: string };
  country: string;

  constructor(data?: JSONObject | Element) {
    super();
    this.createEntity(data);
  }

  public getDisplayName(): string {
    return [this.institution.name, this.address.city, this.country].filter(Boolean).join(', ');
  }

  public clone(): Affiliation {
    const affitiation = new Affiliation();
    affitiation._id = this._id;
    affitiation.label = this.label;
    affitiation.country = this.country;
    affitiation.institution = { name: this.institution.name };
    affitiation.address = { city: this.address.city };
    return affitiation;
  }

  protected fromXML(xml: Element): void {
    this._id = xml.getAttribute('id');
    this.label = getTextContentFromPath(xml, 'label');
    this.institution = {
      name: [
        getTextContentFromPath(xml, 'institution[content-type="dept"]'),
        getTextContentFromPath(xml, 'institution:not([content-type])')
      ]
        .filter(Boolean)
        .join(', ')
    };

    this.address = {
      city: getTextContentFromPath(xml, 'addr-line named-content[content-type="city"]')
    };
    this.country = getTextContentFromPath(xml, 'country');
  }

  protected fromJSON(json: JSONObject): void {
    this._id = json.id as string;
    this.label = json.label as string;
    this.country = json.country as string;
    this.institution = json.institution as { name: string };
    this.address = json.address as { city: string };
  }

  protected createBlank(): void {
    this.label = '';
    this.country = '';
    this.institution = { name: '' };
    this.address = { city: '' };
  }
}

export function createAffiliationsState(affiliationsXml: Element[]): Affiliation[] {
  return affiliationsXml.map((xml) => new Affiliation(xml));
}
