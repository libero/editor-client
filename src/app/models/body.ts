import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { EditorState } from 'prosemirror-state';
import { gapCursor } from 'prosemirror-gapcursor';
import { dropCursor } from 'prosemirror-dropcursor';
import { keymap } from 'prosemirror-keymap';
import { baseKeymap } from 'prosemirror-commands';
import { set } from 'lodash';

import * as bodyConfig from './config/body.config';
import { buildInputRules } from './plugins/input-rules';
import { makeSchemaFromConfig } from 'app/models/utils';
import { PlaceholderPlugin } from 'app/models/plugins/placeholder.plugin';
import { createListKeymap } from 'app/utils/prosemirror/list.helpers';
import { SelectionPlugin } from 'app/models/plugins/selection.plugins';

export function createBodyState(content: Element, id: string): EditorState {
  const schema = makeSchemaFromConfig(bodyConfig.topNode, bodyConfig.nodes, bodyConfig.marks);
  const xmlContentDocument = document.implementation.createDocument('', '', null);

  if (content) {
    xmlContentDocument.appendChild(content);
  }

  set(xmlContentDocument, 'manuscriptPath', `/manuscripts/${id}`);

  return EditorState.create({
    doc: ProseMirrorDOMParser.fromSchema(schema).parse(xmlContentDocument),
    schema,
    plugins: [
      buildInputRules(),
      gapCursor(),
      SelectionPlugin,
      dropCursor({ color: '#0078CF', class: 'drop-cursor' }),
      keymap(createListKeymap(schema)),
      keymap(baseKeymap),
      PlaceholderPlugin('Enter main text')
    ]
  });
}
