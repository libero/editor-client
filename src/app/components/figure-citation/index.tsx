import React from 'react';
import ReactDOM from 'react-dom';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import { EditorView, NodeView } from 'prosemirror-view';
import { ThemeProvider } from '@material-ui/core/styles';

import { theme } from '../../styles/theme';
import { FigureCitationEditorPopup, FiguresListEntry, UNLABELLED_FIGURE_TEXT } from './figure-citation-editor-popup';
import { TextSelection } from 'prosemirror-state';
import { store } from '../../store';
import { Provider } from 'react-redux';

export class FigureCitationNodeView implements NodeView {
  dom?: HTMLAnchorElement;
  contentDOM?: HTMLSpanElement;
  figureEditorContainer: HTMLDivElement;

  constructor(private node: ProsemirrorNode, private view: EditorView, private getPos: () => number) {
    this.dom = document.createElement('a');
    this.dom.style.cursor = 'pointer';
    this.dom.addEventListener('click', this.selectNode);
    this.contentDOM = document.createElement('span');
    this.dom.appendChild(this.contentDOM);
    if (this.node.attrs.figIds === null) {
      this.open();
    }
  }

  selectNode = (): void => {
    this.dom.classList.add('ProseMirror-selectednode');
    this.open();
  };

  deselectNode(): void {
    this.dom.classList.remove('ProseMirror-selectednode');
    this.close();
  }

  update(node: ProsemirrorNode): boolean {
    this.node = node;
    return true;
  }

  open(): void {
    this.figureEditorContainer = this.view.dom.parentNode.appendChild(document.createElement('div'));

    ReactDOM.render(
      <Provider store={store}>
        <ThemeProvider theme={theme}>
          <FigureCitationEditorPopup
            selectedIds={this.node.attrs.figIds || []}
            anchorEl={this.dom}
            onClose={this.close}
            onChange={this.handleChange}
          />
        </ThemeProvider>
      </Provider>,
      this.figureEditorContainer
    );
  }

  stopEvent(): boolean {
    return true;
  }

  ignoreMutation(): boolean {
    return true;
  }

  close = (): void => {
    this.dom.classList.remove('ProseMirror-selectednode');
    if (this.figureEditorContainer) {
      ReactDOM.unmountComponentAtNode(this.figureEditorContainer);
      this.figureEditorContainer.parentNode.removeChild(this.figureEditorContainer);
      this.figureEditorContainer = null;
    }
  };

  handleChange = (figures: FiguresListEntry[]): void => {
    const ids = figures.map(({ id }) => id);
    const change = this.view.state.tr.setNodeMarkup(this.getPos(), null, { figIds: ids });

    if (this.node.textContent === '???' && figures.length > 0) {
      const nodeContent = (figures[0].name || UNLABELLED_FIGURE_TEXT).replace(/\.$/, '');
      change
        .setSelection(TextSelection.create(change.doc, this.getPos() + 1, this.getPos() + this.node.nodeSize - 1))
        .replaceSelectionWith(this.view.state.schema.text(nodeContent));
    } else if (figures.length === 0) {
      change
        .setSelection(TextSelection.create(change.doc, this.getPos() + 1, this.getPos() + this.node.nodeSize - 1))
        .replaceSelectionWith(this.view.state.schema.text('???'));
    }

    this.view.dispatch(change);
  };
}
