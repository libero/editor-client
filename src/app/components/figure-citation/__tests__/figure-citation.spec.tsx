import React from 'react';
import { Node as ProsemirrorNode } from 'prosemirror-model';
import { EditorView } from 'prosemirror-view';
import { EditorState } from 'prosemirror-state';
import { ReplaceAroundStep } from 'prosemirror-transform';

import { FigureCitationNodeView } from 'app/components/figure-citation/index';
import { createBodyState } from 'app/models/body';

jest.mock('prosemirror-view');
jest.mock('app/components/figure-citation/figure-citation-editor-popup', () => ({
  FigureCitationEditorPopup: () => <div data-cmp="FigureCitationEditorPopup" />
}));
jest.mock('@material-ui/core/styles', () => {
  return {
    ThemeProvider: ({ children }) => <>{children}</>,
    createMuiTheme: jest.fn(),
    makeStyles: jest.requireActual('@material-ui/core/styles').makeStyles
  };
});

global['document']['createRange'] = () => ({
  setStart: jest.fn(),
  setEnd: jest.fn(),
  commonAncestorContainer: {
    nodeName: 'BODY',
    ownerDocument: document
  }
});

describe('Figure citation node view', () => {
  it('should render node', () => {
    const [state, node] = givenState();
    const view = givenEditorView(state);
    const nodeView = new FigureCitationNodeView(node, view, jest.fn());
    expect(nodeView.dom.innerHTML).toMatchSnapshot();
  });

  it('should open editor on click', () => {
    const parentContainer = document.createElement('div');
    parentContainer.id = 'view-parent-container';
    const [state, node] = givenState();
    const view = givenEditorView(state);
    const nodeView = new FigureCitationNodeView(node, view, jest.fn());
    parentContainer.appendChild(view.dom);
    nodeView.dom.dispatchEvent(new Event('click'));
    expect(parentContainer.innerHTML).toMatchSnapshot();
  });

  it('should update node markup', () => {
    const [state, node] = givenState();
    const view = givenEditorView(state);
    const nodeView = new FigureCitationNodeView(node, view, () => 1);
    nodeView.handleChange([
      { id: 'id1', name: 'figure 1' },
      { id: 'id2', name: 'figure 2' }
    ]);
    expect(view.dispatch['mock'].calls[0][0].steps[0]).toBeInstanceOf(ReplaceAroundStep);
  });

  it('should remove popup on close', () => {
    const parentContainer = document.createElement('div');
    parentContainer.id = 'view-parent-container';
    const [state, node] = givenState();
    const view = givenEditorView(state);
    const nodeView = new FigureCitationNodeView(node, view, jest.fn());
    parentContainer.appendChild(view.dom);
    nodeView.dom.dispatchEvent(new Event('click'));
    nodeView.close();
    expect(parentContainer.innerHTML).toBe(view.dom.outerHTML);
  });
});

function givenState(): [EditorState, ProsemirrorNode] {
  let state = createBodyState(document.createElement('div'), '');
  const node = state.schema.nodes.figureCitation.createAndFill({ figIds: ['id1'] }, state.schema.text('test content'));
  state = state.apply(state.tr.insert(0, node));
  return [state, node];
}

function givenEditorView(state: EditorState): EditorView {
  const dom = document.createElement('div');
  dom.id = 'view-container';
  const view = new EditorView(dom, {
    state: null,
    dispatchTransaction: jest.fn()
  });
  view.dom = dom;
  view.state = state;
  view.dispatch = jest.fn();
  return view;
}
