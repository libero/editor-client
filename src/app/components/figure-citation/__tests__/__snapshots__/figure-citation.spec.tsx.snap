// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Figure citation node view should open editor on click 1`] = `"<div id=\\"view-container\\"></div><div></div>"`;

exports[`Figure citation node view should render node 1`] = `"<span></span>"`;
